cocos/gfx/webgpu/define.ts

```diff
@@ -0,0 +1,60 @@
+import { DescriptorSet } from '../base/descriptor-set';
+import { DescriptorSetLayout } from '../base/descriptor-set-layout';
+import { WebGPUBuffer } from './webgpu-buffer';
+import { WebGPUDevice } from './webgpu-device';
+import { WebGPUSampler } from './webgpu-sampler';
+import { WebGPUTexture } from './webgpu-texture';
+
+export class WebGPUDeviceManager {
+    static get instance (): WebGPUDevice {
+        return WebGPUDeviceManager._instance!;
+    }
+    static setInstance (instance: WebGPUDevice): void {
+        WebGPUDeviceManager._instance = instance;
+    }
+    private static _instance: WebGPUDevice | null = null;
+}
+
+function hashCombine (hash, currHash: number): number {
+    return currHash ^= (hash >>> 0) + 0x9e3779b9 + (currHash << 6) + (currHash >> 2);
+}
+
+export function hashCombineNum (val: number, currHash: number): number {
+    const hash = 5381;
+    return hashCombine((hash * 33) ^ val, currHash);
+}
+
+export function hashCombineStr (str: string, currHash: number): number {
+    // DJB2 HASH
+    let hash = 5381;
+    for (let i = 0; i < str.length; i++) {
+        hash = (hash * 33) ^ str.charCodeAt(i);
+    }
+    return hashCombine(hash, currHash);
+}
+
+export class DefaultResources {
+    // hash, targetResource
+    buffersDescLayout: Map<number, WebGPUBuffer> = new Map<number, WebGPUBuffer>();
+    texturesDescLayout: Map<number, WebGPUTexture> = new Map<number, WebGPUTexture>();
+    samplersDescLayout: Map<number, WebGPUSampler> = new Map<number, WebGPUSampler>();
+    buffer!: WebGPUBuffer;
+    texture!: WebGPUTexture;
+    sampler!: WebGPUSampler;
+    setLayout!: DescriptorSetLayout;
+    descSet!: DescriptorSet;
+}
+
+export function isBind(binds: number[], compares: number[]): boolean {
+    return binds.length === compares.length && binds.every(bind => compares.includes(bind));
+}
+
+export function copyNumbersToTarget(source: number[], target: number[], start: number, count: number): void {
+    const sliceToCopy = source.slice(start, start + count);
```

Comment:
Should check range first? The count may exceed source or target range.

Reply:
Already modified